ui/src/app/job-details/resources/resources.component.ts

```diff
@@ -0,0 +1,106 @@
+import {
+  Component,
+  Input,
+  OnChanges,
+  SimpleChanges,
+} from '@angular/core';
+import {
+  MatExpansionPanel,
+  MatTabChangeEvent
+} from '@angular/material'
+
+import {JobMetadataResponse} from '../../shared/model/JobMetadataResponse';
+import {ResourceUtils} from '../../shared/utils/resource-utils';
+import {GcsService} from '../../core/gcs.service';
+
+@Component({
+  selector: 'jm-resources',
+  templateUrl: './resources.component.html',
+  styleUrls: ['./resources.component.css'],
+})
+export class JobResourcesComponent implements OnChanges {
+  @Input() job: JobMetadataResponse;
+
+  script: string = '';
+  inputs: Array<string> = [];
+  logFileData: Map<string, string> = new Map();
+  logFiles: Array<string> = [];
+  outputs: Array<string> = [];
+  currentResourcesTab: string = '';
+
+  constructor(private readonly gcsService: GcsService) {}
+
+  ngOnChanges(changes: SimpleChanges) {
+    this.job = changes.job.currentValue;
+    if (this.job.extensions) {
+      if (this.job.extensions.logs) {
+        this.logFiles = Object.keys(this.job.extensions.logs).sort();
+        for (let file of Object.keys(this.job.extensions.logs)) {
+          let bucket = ResourceUtils.getResourceBucket(this.job.extensions.logs[file]);
+          let object = ResourceUtils.getResourceObject(this.job.extensions.logs[file]);
+          this.gcsService.readObject(bucket, object).then(data => {
+            if (data.length > 0) {
+              this.logFileData.set(file, data);
+            } else {
+              // Remove any files from the tab which are completely empty
+              this.logFiles.splice(this.logFiles.indexOf(file), 1);
+            }
+          })
+        }
+      }
+
+      if (this.job.extensions.script) {
+        this.script = this.job.extensions.script;
+      }
+    }
+
+    if (this.job.inputs) {
+      this.inputs = Object.keys(this.job.inputs).sort();
+    }
+    if (this.job.outputs) {
+      this.outputs = Object.keys(this.job.outputs).sort();
+    }
+
+    if (this.hasInputs()) {
+      this.currentResourcesTab = 'Inputs';
+    } else if (this.hasOutputs()) {
+      this.currentResourcesTab = 'Outputs';
+    } else if (this.hasScript()) {
+      this.currentResourcesTab = 'Script';
+    } else if (this.logFiles.length > 0) {
+      this.currentResourcesTab = this.logFiles[0];
+    }
+  }
+
+  hasInputs(): boolean {
+    return this.inputs.length > 0;
+  }
+
+  hasOutputs(): boolean {
+    return this.outputs.length > 0;
+  }
+
+  hasScript(): boolean {
+    return this.job.extensions && !!this.job.extensions.script;
+  }
+
+  keepExpanded(): boolean {
+    return !this.job.extensions || !this.job.extensions.tasks;
+  }
+
+  tabChanged(event: MatTabChangeEvent) {
+    this.currentResourcesTab = event.tab.textLabel;
```

Comment:
Is there something you could use in lieu of the display text as the identifier here? If a backend returned "Script" or any of the other hardcoded tabs as a log file name, there would be a collision here.

Reply:
Really good point, its a bit tough since the tabs are dynamic based on what data is present so we can't just use an index. I came up with a solution I'm okay with which basically gives each tab an "ID" mapped to its title. For top-level fields this is the same as the name, and for log files it is prefixed with `"log-"`.

Let me know what you think.

Reply:
Hmmm, I think this approach still has the same issue, since you are keying the map by the tab name.

Perhaps array of tab "ID" is the right approach. Noting that you'd need to also splice out log data if they fail to load. currentResourcesTab would then have the tab ID value